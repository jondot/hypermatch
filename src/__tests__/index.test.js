const run = require('../')

const t = [
  'or',
  ['or', ['includes', 'name', ['a']], ['equals', 'browser', 'firefox']],
  ['includes', 'uid', ['user1']]
]

const t1 = [
  'and',
  ['includes', 'name', ['a']],
  ['equals', 'browser', 'firefox']
]

const bad = [
  'and',
  ['incluuudes', 'name', ['a']],
  ['equals', 'browser', 'firefox']
]

const t3 = ['not', ['equals', 'browser', 'firefox']]

const r1 = ['and', ['equals', 'uid', '1'], ['regex', 'name', /moo.*/]]

class Tracer {
  constructor() {
    this.actions = []
    this.log = this.log.bind(this)
  }
  log(...args) {
    this.actions.push(args)
  }
}

const testRun = (res, tree, props) => {
  test(`${JSON.stringify(tree)} ${res ? '✓' : '✕'} ${JSON.stringify(
    props
  )}`, () => {
    const tracer = new Tracer()
    expect(run(tree, props, { trace: true, log: tracer.log })).toEqual(res)
    expect(tracer.actions).toMatchSnapshot()
  })
}

test('handles bad syntax', () =>
  expect(() => {
    run(bad, { name: 'a' })
  }).toThrow())

testRun(false, t1, { uid: 'doesnt', name: 'match' })
testRun(true, t1, { name: 'a', browser: 'firefox' })
testRun(false, t1, { name: 'a', browser: 'nope!' })

testRun(true, t, { uid: 'user1', name: 'b', browser: 'android' })

testRun(true, t3, { browser: 'android' })
testRun(false, t3, { browser: 'firefox' })

testRun(true, ['equals', 'name', 'foobar'], { name: 'foobar' })
testRun(false, ['equals', 'name', 'foobar'], { name: 'foo' })

testRun(true, ['includes', 'name', ['m1', 'm2']], { name: 'm2' })
testRun(false, ['includes', 'name', ['m1', 'm2']], { name: 'm3' })

testRun(false, t, { name: 'bad', browser: 'foobar', uid: 'poop' })

// regex
testRun(true, ['regex', 'name', 'moo.*'], { name: 'moomoo moo!', uid: '1' })
testRun(true, r1, { name: 'moomoo moo!', uid: '1' })
testRun(false, r1, { name: 'poop', uid: '1' })

// subset
testRun(true, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3', 'O-4']], {
  orderIds: ['O-1', 'O-2', 'O-3']
})
testRun(false, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3']], {
  orderIds: ['O-2', 'O-3', 'O-4']
})
testRun(true, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3']], { orderIds: [] })
testRun(true, ['subset', 'orderIds', []], { orderIds: [] })

// intersects
testRun(true, ['intersects', 'professions', ['barista', 'baker', 'bartender', 'waiter']], {
  professions: ['chef', 'baker', 'bartender']
})
testRun(false, ['intersects', 'professions', ['baker', 'barista']], { 'professions': ['waiter'] })
testRun(true, ['intersects', 'professions', ['barista']], { professions: ['barista'] })

// range
testRun(true, ['range', 'age', [18, 40]], { age: 23 })
testRun(false, ['range', 'age', [18, 40]], { age: 66 })

// exists
testRun(true, ['exists', 'age'], { age: 23 })
testRun(false, ['exists', 'age'], { foobar: 23 })

// lt, gt
testRun(false, ['lt', 'age', 12], { age: 23 })
testRun(true, ['gt', 'age', 12], { age: 23 })
testRun(true, ['lt', 'age', 40], { age: 23 })
testRun(false, ['gt', 'age', 40], { age: 23 })
