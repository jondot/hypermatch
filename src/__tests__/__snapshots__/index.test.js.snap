// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`["all","users",["and",["regex","user","^john"],["exists","active"]]] ✓ {"users":[{"user":"john","age":36,"active":true},{"user":"johnathan","age":40,"active":true},{"user":"johny","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "all",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": true,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": true,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["all","users",["and",["regex","user","^john"],["exists","active"]]] ✕ {"users":[{"user":"john","age":36,"active":true},{"user":"johnathan","age":40,"active":false},{"user":"johny","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "all",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": true,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["and",["equals","uid","1"],["regex","name",{}]] ✓ {"name":"moomoo moo!","uid":"1"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "equals",
        "uid",
        "1",
      ],
      Array [
        "regex",
        "name",
        /moo\\.\\*/,
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "1",
      "1",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "regex",
    Array [
      /moo\\.\\*/,
      "moomoo moo!",
    ],
  ],
]
`;

exports[`["and",["equals","uid","1"],["regex","name",{}]] ✕ {"name":"poop","uid":"1"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "equals",
        "uid",
        "1",
      ],
      Array [
        "regex",
        "name",
        /moo\\.\\*/,
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "1",
      "1",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "regex",
    Array [
      /moo\\.\\*/,
      "poop",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✓ {"name":"a","browser":"firefox"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "a",
      ],
      "a",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "firefox",
      "firefox",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✕ {"name":"a","browser":"nope!"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "a",
      ],
      "a",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "nope!",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✕ {"uid":"doesnt","name":"match"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "match",
    ],
  ],
]
`;

exports[`["any","users",["and",["regex","user","^john"],["exists","active"]]] ✓ {"users":[{"user":"john","age":36,"active":false},{"user":"johnathan","age":40,"active":false},{"user":"johny","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "any",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": false,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["any","users",["and",["regex","user","^john"],["exists","active"]]] ✕ {"users":[{"user":"john","age":36,"active":false},{"user":"johnathan","age":40,"active":false},{"user":"johny","age":21,"active":false}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "any",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": false,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": false,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["equals","name","foobar"] ✓ {"name":"foobar"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "foobar",
      "foobar",
    ],
  ],
]
`;

exports[`["equals","name","foobar"] ✕ {"name":"foo"} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "foobar",
      "foo",
    ],
  ],
]
`;

exports[`["exists","age"] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "exists",
    Array [
      23,
    ],
  ],
]
`;

exports[`["exists","age"] ✕ {"foobar":23} 1`] = `Array []`;

exports[`["gt","age",12] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "gt",
    Array [
      12,
      23,
    ],
  ],
]
`;

exports[`["gt","age",40] ✕ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "gt",
    Array [
      40,
      23,
    ],
  ],
]
`;

exports[`["includes","name",["m1","m2"]] ✓ {"name":"m2"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "m1",
        "m2",
      ],
      "m2",
    ],
  ],
]
`;

exports[`["includes","name",["m1","m2"]] ✕ {"name":"m3"} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "m1",
        "m2",
      ],
      "m3",
    ],
  ],
]
`;

exports[`["intersects","professions",["baker","barista"]] ✕ {"professions":["waiter"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "intersects",
    Array [
      Array [
        "baker",
        "barista",
      ],
      Array [
        "waiter",
      ],
    ],
  ],
]
`;

exports[`["intersects","professions",["barista","baker","bartender","waiter"]] ✓ {"professions":["chef","baker","bartender"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "intersects",
    Array [
      Array [
        "barista",
        "baker",
        "bartender",
        "waiter",
      ],
      Array [
        "chef",
        "baker",
        "bartender",
      ],
    ],
  ],
]
`;

exports[`["intersects","professions",["barista"]] ✓ {"professions":["barista"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "intersects",
    Array [
      Array [
        "barista",
      ],
      Array [
        "barista",
      ],
    ],
  ],
]
`;

exports[`["lt","age",12] ✕ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "lt",
    Array [
      12,
      23,
    ],
  ],
]
`;

exports[`["lt","age",40] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "lt",
    Array [
      40,
      23,
    ],
  ],
]
`;

exports[`["none","users",["and",["regex","user","^john"],["exists","active"]]] ✓ {"users":[{"user":"bob","age":36,"active":false},{"user":"bill","age":40,"active":false},{"user":"max","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "none",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": false,
          "age": 36,
          "user": "bob",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "bill",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "max",
        },
      ],
    ],
  ],
]
`;

exports[`["none","users",["and",["regex","user","^john"],["exists","active"]]] ✕ {"users":[{"user":"bob","age":36,"active":false},{"user":"johnathan","age":40,"active":true},{"user":"max","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "none",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": false,
          "age": 36,
          "user": "bob",
        },
        Object {
          "active": true,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "max",
        },
      ],
    ],
  ],
]
`;

exports[`["not",["equals","browser","firefox"]] ✓ {"browser":"android"} 1`] = `
Array [
  Array [
    "op[1]  ",
    "not",
    Array [
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "android",
    ],
  ],
]
`;

exports[`["not",["equals","browser","firefox"]] ✕ {"browser":"firefox"} 1`] = `
Array [
  Array [
    "op[1]  ",
    "not",
    Array [
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "firefox",
      "firefox",
    ],
  ],
]
`;

exports[`["one","users",["and",["regex","user","^john"],["exists","active"]]] ✓ {"users":[{"user":"john","age":36,"active":false},{"user":"johnathan","age":40,"active":false},{"user":"johny","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "one",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": false,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["one","users",["and",["regex","user","^john"],["exists","active"]]] ✕ {"users":[{"user":"john","age":36,"active":true},{"user":"johnathan","age":40,"active":false},{"user":"johny","age":21,"active":true}]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "one",
    Array [
      Array [
        "and",
        Array [
          "regex",
          "user",
          "^john",
        ],
        Array [
          "exists",
          "active",
        ],
      ],
      Array [
        Object {
          "active": true,
          "age": 36,
          "user": "john",
        },
        Object {
          "active": false,
          "age": 40,
          "user": "johnathan",
        },
        Object {
          "active": true,
          "age": 21,
          "user": "johny",
        },
      ],
    ],
  ],
]
`;

exports[`["or",["or",["includes","name",["a"]],["equals","browser","firefox"]],["includes","uid",["user1"]]] ✓ {"uid":"user1","name":"b","browser":"android"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "or",
        Array [
          "includes",
          "name",
          Array [
            "a",
          ],
        ],
        Array [
          "equals",
          "browser",
          "firefox",
        ],
      ],
      Array [
        "includes",
        "uid",
        Array [
          "user1",
        ],
      ],
    ],
  ],
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "b",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "android",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "user1",
      ],
      "user1",
    ],
  ],
]
`;

exports[`["or",["or",["includes","name",["a"]],["equals","browser","firefox"]],["includes","uid",["user1"]]] ✕ {"name":"bad","browser":"foobar","uid":"poop"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "or",
        Array [
          "includes",
          "name",
          Array [
            "a",
          ],
        ],
        Array [
          "equals",
          "browser",
          "firefox",
        ],
      ],
      Array [
        "includes",
        "uid",
        Array [
          "user1",
        ],
      ],
    ],
  ],
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "bad",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "foobar",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "user1",
      ],
      "poop",
    ],
  ],
]
`;

exports[`["range","age",[18,40]] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "range",
    Array [
      Array [
        18,
        40,
      ],
      23,
    ],
  ],
]
`;

exports[`["range","age",[18,40]] ✕ {"age":66} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "range",
    Array [
      Array [
        18,
        40,
      ],
      66,
    ],
  ],
]
`;

exports[`["regex","name","moo.*"] ✓ {"name":"moomoo moo!","uid":"1"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "regex",
    Array [
      "moo.*",
      "moomoo moo!",
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3","O-4"]] ✓ {"orderIds":["O-1","O-2","O-3"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
        "O-4",
      ],
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3"]] ✓ {"orderIds":[]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
      Array [],
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3"]] ✕ {"orderIds":["O-2","O-3","O-4"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
      Array [
        "O-2",
        "O-3",
        "O-4",
      ],
    ],
  ],
]
`;

exports[`["subset","orderIds",[]] ✓ {"orderIds":[]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [],
      Array [],
    ],
  ],
]
`;
